src/plugins/vdo_stats.c

```diff
@@ -132,76 +133,127 @@ static void add_computed_stats (GHashTable *stats) {
+        g_set_error (error, BD_LVM_ERROR, BD_LVM_ERROR_DM_ERROR,
+                     "Failed to set message for DM task");
+        dm_task_destroy (dmt);
+        return NULL;
+    }
 
-G_GNUC_INTERNAL GHashTable *
-vdo_get_stats_full (const gchar *name, GError **error) {
-    GHashTable *stats;
-    GDir *dir;
-    gchar *stats_dir;
-    const gchar *direntry;
-    gchar *s;
-    gchar *val = NULL;
-    g_autofree gchar *dm_node = NULL;
-    GError *l_error = NULL;
-
-    /* try "new" (kvdo >= 8) path first -- /sys/block/dm-X/vdo/statistics */
-    dm_node = _dm_node_from_name (name, error);
-    if (dm_node == NULL) {
-        g_prefix_error (error, "Failed to get DM node for %s: ", name);
+    if (!dm_task_run (dmt)) {
+        g_set_error (error, BD_LVM_ERROR, BD_LVM_ERROR_DM_ERROR,
+                     "Failed to run DM task");
+        dm_task_destroy (dmt);
+        return NULL;
+    }
+
+    response = dm_task_get_message_response (dmt);
+    if (!response) {
+        g_set_error (error, BD_LVM_ERROR, BD_LVM_ERROR_DM_ERROR,
+                     "Failed to get response from the DM task");
+        dm_task_destroy (dmt);
         return NULL;
     }
 
-    stats_dir = g_build_path (G_DIR_SEPARATOR_S, "/sys/block", dm_node, "vdo/statistics", NULL);
-    dir = g_dir_open (stats_dir, 0, &l_error);
-    if (dir == NULL) {
-        bd_utils_log_format (BD_UTILS_LOG_INFO,
-                             "Failed to read VDO stats using the new API, falling back to %s: %s",
-                             VDO_SYS_PATH, l_error->message);
-        g_free (stats_dir);
-        g_clear_error (&l_error);
-
-        /* lets try /sys/kvdo */
-        stats_dir = g_build_path (G_DIR_SEPARATOR_S, VDO_SYS_PATH, name, "statistics", NULL);
-        dir = g_dir_open (stats_dir, 0, error);
-        if (dir == NULL) {
-            g_prefix_error (error, "Error reading statistics from %s: ", stats_dir);
-            g_free (stats_dir);
-            return NULL;
-        }
+    if (!yaml_parser_initialize (&parser)) {
+        g_set_error (error, BD_LVM_ERROR, BD_LVM_ERROR_DM_ERROR,
+                     "Failed to get initialize YAML parser");
+        dm_task_destroy (dmt);
+        return NULL;
     }
 
     stats = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
-    while ((direntry = g_dir_read_name (dir))) {
-        s = g_build_filename (stats_dir, direntry, NULL);
-        if (! g_file_get_contents (s, &val, NULL, error)) {
-            g_prefix_error (error, "Error reading statistics from %s: ", s);
-            g_free (s);
-            g_hash_table_destroy (stats);
-            stats = NULL;
-            break;
-        }
-        g_hash_table_replace (stats, g_strdup (direntry), g_strdup (g_strstrip (val)));
-        g_free (val);
-        g_free (s);
-    }
-    g_dir_close (dir);
-    g_free (stats_dir);
+    yaml_parser_set_input_string (&parser, (guchar *) response, strlen (response));
```

Comment:
Who the hell invented putting yaml-formatted string in a dm message??!

Reply:
The original output of VDO stats was also in yaml (but formatted differently) so maybe they wanted to keep it? But definitely unexpected.